Updates stylelint, eslint and dependencies

osk opened the pull request:
Updates eslint to 4.7 and other dependencies to latest

eslint throws `DeprecationWarning: [eslint] The 'ecmaFeatures' config file property is deprecated, and has no effect` for some rules that airbnb relies on

osk committed:
Add `overflow`, `order` and grid properties to stylelint config

osk committed:
Update stylelint & dependencies

osk committed:
Update eslint & dependencies

osk committed:
Remove deprecated "ecmaFeatures" key in eslint config

osk commented:
Looks like `xo` [hasn't updated to eslint 4.x](https://github.com/sindresorhus/xo/issues/230) so there are issues with `import/*` rules that cause the tests to fail.

gakimball commented on packages/eslint-config/index.js:
```diff
@@ -48,7 +44,7 @@ module.exports = {
     'no-param-reassign': ['error', {
       props: false
-    }]
+    }],
```
Drop this comma to fix one of the linting errors. The rest of the linting errors are from those issues with `import` I believe.

(xo has some weird rules, I mostly use it for small projects because you don't have to set anything up.)

osk replied:
This becomes unneeded since I added keys after

gakimball commented:
Looks good to me 👍

gakimball commented:
There's a few rules in the newer versions of `eslint-plugin-react` that I'd like to propose adding (that aren't in Airbnb's config):

- `no-unused-state`: state variables must be used in a component. Useful when you're removing functionality from a component to keep things clean.
- `default-props-match-prop-types` the items in `propTypes` and `defaultProps` must match. Once again, useful for when you're cleaning things up so you don't miss anything.
- `no-typos`: catches you if you misspell a PropType, like `PropTypes.fun` or `PropTypes.boo`.

These are all issues I've run into before when combing through code. They're small things but they help keep code clean 👌 

@finnursig @osk thoughts?

gakimball commented:
Also, using something like Greenkeeper to alert us when all these libraries go out of date would be pretty useful.

osk committed:
Add new react eslint rules

osk commented:
Yeah, these seem like good rules to add, so I added them.

gakimball commented:
One more suggestion: let's move `babel-eslint` from `peerDependencies` to `dependencies`. The ability to independently update the eslint parser won't be too terribly useful.

The only reason it wasn't in there initially is because... I forgot to add it 😅

osk committed:
Update stylelint, eslint and eslint-config-airbnb deps

eslint-config-airbnb has updated to 4.x of eslint :tada:

osk commented:
Good thing I didn't merge this because eslint-config-airbnb was updated to eslint 4.x! They also updated all dependencies so we're not getting warnings because of peers 😊

Part of the update was to [enable all the new react rules](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/CHANGELOG.md#1600--2017-10-06), so I removed them again and tweaked some new rules by running against the starter-kit, see [PR](https://github.com/ueno-llc/starter-kit-universally/pull/62) that can be updated after this is merged.
